pkg/trace/stats/sublayers.go

```diff
@@ -34,6 +43,181 @@ func (v SublayerValue) GoString() string {
+// for a trace with max n spans
+func (c *SublayerCalculator) reset(n int) {
+	if c.maxSpans < n {
+		c.initFields(n)
+	}
+	for i := 0; i < n; i++ {
+		c.activeSpansIndex[i] = inactiveSpan
+		c.activeSpans[i] = 0
+		c.openSpans[i] = false
+		c.nChildren[i] = 0
+		c.execDuration[i] = 0
+	}
+}
+
+// a timestamp is the start or end of a span
+// we store the span and parent index for each timestamp instead
+// of storing a pointer to the span to be able to reuse the sublayer calculator arrays
+type timestamp struct {
+	spanStart bool
+	spanIdx   int
+	parentIdx int
+	ts        int64
+}
+
+// timestampArray is an array of timestamp. Used to define Len, Swap and Less to be able to sort the array
+type timestampArray []timestamp
+
+func (t timestampArray) Len() int      { return len(t) }
+func (t timestampArray) Swap(i, j int) { t[i], t[j] = t[j], t[i] }
+
+// for spans with a duration of 0, we need to open them before closing them
+func (t timestampArray) Less(i, j int) bool {
+	return t[i].ts < t[j].ts || (t[i].ts == t[j].ts && t[i].spanStart && !t[j].spanStart)
+}
+
+// buildTimestamps builds puts starts and ends of each span in an array of timestamps, and sorts the array.
+// result if stored in the SublayerCalculator
+func (c *SublayerCalculator) buildTimestamps(trace pb.Trace) {
+	for i, span := range trace {
+		c.timestamps[2*i] = timestamp{spanStart: true, spanIdx: i, parentIdx: c.parentIdx[i], ts: span.Start}
+		c.timestamps[2*i+1] = timestamp{spanStart: false, spanIdx: i, parentIdx: c.parentIdx[i], ts: span.Start + span.Duration}
+	}
+	sort.Sort(c.timestamps[:2*len(trace)])
+}
+
+// computeParentIdx builds the mapping spanIdx --> parentIdx
+// it stores the result in the SublayerCalculator
+func (c *SublayerCalculator) computeParentIdx(trace pb.Trace) {
+	idToIdx := make(map[uint64]int, len(trace))
+	for idx, span := range trace {
+		idToIdx[span.SpanID] = idx
+	}
+	for idx, span := range trace {
+		if parentIdx, ok := idToIdx[span.ParentID]; ok {
+			c.parentIdx[idx] = parentIdx
+		} else {
+			c.parentIdx[idx] = unknownSpan
+		}
+	}
+}
+
+// computeExecDuration computes the exec duration of each span in the trace
+//
+// the algorithm does a traversal of all ordered timestamps (start and end of each span) in the trace
+// Let's define a few terms to make the algorithm clearer.
+// For a given timestamp:
+//
+// Open Span: The Span has started before this timestamp, and ends after this timestamp.
+// Number of Children of a Span: At a given timestamp, number of direct children spans that are currently open
+// Active Span: a Span is active at a given timestamp, if it is open and if it has no open children at the timestamp
+// Exec duration: for each period where the span is active, its Exec Duration increases by deltaT/numberOfActiveSpans
+//     so if a span is active during 20ms, in parallel with an other span active at the same time, its exec duration will be 10ms
```

Comment:
This part should go above each field in the `SublayerCalculator` so that the reader can always associate it with its field name.